Automatically link source snippets in the documentation to the code

Vampire opened the pull request:
Will look like this when produced:

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/tdLND5eJpRKUT3xAEJML/918cb757-86d9-419d-a61e-ea646ef6e370.png)

And additionally logs warnings for listing blocks that do not have a source include like
> WARNUNG: listing with only inline code found; if this is not source from a file, consider using a literal block instead; first line: foo.bar() >> { throw new IOException() } >>> [1, 2, 3] >> { throw new RuntimeException() }

Vampire commented:
* **#1904** <a href="https://app.graphite.dev/github/pr/spockframework/spock/1904?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Vampire and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

leonard84 commented:
Nice, I'll review it in detail later. However, if you already have this, then you could easily add the Groovy Web Console integration, see `From a repository` in https://gwc-experiment.appspot.com/

leonard84 commented on build-logic/asciidoc-extensions/src/main/java/org/spockframework/plugins/asciidoctor/IncludedSourceLinker.java:
```diff
@@ -0,0 +1,257 @@
+import static org.asciidoctor.extension.Contexts.LITERAL;
+import static org.asciidoctor.log.Severity.WARN;
+
+public class IncludedSourceLinker {
```
This is quite a lot more complex than the other processor, can we reasonably write unit tests for this?

Vampire replied:
Maybe. I mainly hacked this together quickly and wanted at least first feedback if you like it at all. :-D
Also I hope to get rid of half of the code, when Asciidoctor hopefully exposes a way to reuse the tag filterting logic.
The only way to get the filename of the included file is a custom include processor and with a custom include processor you currently have to do all the include handling yourself.

leonard84 replied:
Is there an Asciidoctor issue for this?

Vampire replied:
It indeed was hacked together too quickly and had some problems.
I now did a 1:1 port from the original Ruby code in Asciidoctor to behave exactly the same as of now,
and also added `lines` and `indent` support on the include directive.
Now only `leveloffset` which should not make sense, `encoding` which should not be necessary, and `opts` which likely isn't needed are unsupported with the custom include handling.

As most of the code indeed is the 1:1 port from the Asciidoctor Ruby code with added code comments, that hopefully can be thrown out at some point, I'm not sure whether I really want to write tests for it. :-D

> Is there an Asciidoctor issue for this?

Yes: asciidoctor/asciidoctor#571

leonard84 commented on build-logic/asciidoc-extensions/src/main/java/org/spockframework/plugins/asciidoctor/IncludedSourceLinker.java:
```diff
@@ -0,0 +1,257 @@
+      }
+
+      @Override
+      public void process(Document document, PreprocessorReader reader, String target, Map<String, Object> attributes) {
```
This method is quite long. Can you break it up in descriptively named methods? Or, at least add section comments about what each section does.

Vampire commented:
> if you already have this, then you could easily add the Groovy Web Console integration, see `From a repository` in https://gwc-experiment.appspot.com/ this only works for a commit-hash and not a branch reference though.

Indeed, but I have some concerns:
- I still hope you can reactivate meetspock.appspot.com
- You obviously too as it is still the one mentioned in the docs
- `gwc-experiment` is just an experiment, isn't it? If we really want to promote it as a replacement, it should be renamed
- It would be nice if it also worked with a branch name. The release docs will have the commit hash in there anyway, but for example when generating the docs locally, it would be nice if it worked too.

Vampire commented:
Also, often those specs do not work. For example if it extends `EmbeddedSpecification`, it cannot execute as this class is only available in our tests.

leonard84 commented:
> I still hope you can reactivate meetspock.appspot.com
> You obviously too as it is still the one mentioned in the docs

Hm, I thought I had removed the references, I replaced it in the .

> `gwc-experiment` is just an experiment, isn't it? If we really want to promote it as a replacement, it should be renamed

It's not an experiment anymore. It is just that we couldn't agree on a domain name and fund it's purchase and maintenance. /cc @glaforge

> It would be nice if it also worked with a branch name. The release docs will have the commit hash in there anyway, but for example when generating the docs locally, it would be nice if it worked too.

I agree, but it is a limitation of the [service](https://raw.githack.com/) we use to be able to load the raw git content I don't really know the limitation. 

> Also, often those specs do not work. For example if it extends `EmbeddedSpecification`, it cannot execute as this class is only available in our tests.

Yes, that is true, would it be possible to have an option for the codeblock/include to enable or disable it.

Vampire commented:
> Hm, I thought I had removed the references, I replaced it in the .

You only replaced it in the readme, but the docs still have meetspock in two places in the `getting_started.adoc`.

> It's not an experiment anymore. It is just that we couldn't agree on a domain name and fund it's purchase and maintenance.

Then at least name it `gwc.appspot.com` or something?
I always refrained from using it or let alone recommending it due to it being an experiment and I'm sure others feel the same.

> I agree, but it is a limitation of the [service](https://raw.githack.com/) we use to be able to load the raw git content I don't really know the limitation.

Hm, I see. It caches by URL and serves from CF's CDN. This is hardly possible for a URL with changing content.
So you would need to directly get the raw file from GitHub and then risk to get rate-limited. :-/
Actually, it also does not work for tags or for abbreviated commit IDs, which both should be considered non-changing.
It only works for full-blown commit IDs.
Maybe a viable option to at least support branches and tags would be to resolve the name to ID yourself with something like
```bash
git ls-remote https://github.com/spockframework/spock.git refs/heads/<ref> refs/tags/<ref>
```
That would give you the current commit ID on the remote server for the branch and / or tag with that name.
Or just
```bash
git ls-remote https://github.com/spockframework/spock.git
```
and then filtering the list yourself which would work better for multi-part branch names like `vampire/auto-link-source-snippets`.

We could also just use `code` or `codez` attribute and directly supply the source code, but then the links in the documentation will probably become quite large. Unless of course we only contain the shown snippet. But then it will never compile in most cases in our docs I guess.

Anyway, all for the play button is in place already now.
It is just not effective, as the document attribute `gwc-base` is not set.
As soon as it is set, play icons are added to all listing blocks where the first include comes from a Groovy file.
If the link to the source file does not contain `/master/`, the play button will also link to the `gwc-base` accordingly.

If you think we should go with the experiment, I just need to add that one document attribute and the play buttons will be in place:
![image](https://github.com/spockframework/spock/assets/325196/0134424d-7f62-4f87-b958-86711c9ca8d3)

> Yes, that is true, would it be possible to have an option for the codeblock/include to enable or disable it.

I added a boolean attribute named `play` to the enclosing block. If it is set to a non-`true` value, the play icon will not be shown.
I'm just not sure whether you really want to use that.
The idea was, that all this happens automagically, without the need to manually do something different to normal AsciiDoc.
I don't now whether it is better to show the icon to view the code and then fail if it does not work, or whether it is better to not show the play button for unknown reason (from the User's perspective).

I think if we show it, we should just always show it.
You will also have the problem that old docs might refer to files that are no longer available,
or milestone does refer to files that are not yet available in the gwc-included Spock version, ...

leonard84 commented:
When running `:asciidoctor` with these changes, I get a lot of these errors:

```
Converting /Users/leonard/ghq/github.com/spockframework/spock/docs/data_driven_testing.adoc
Mar 17, 2024 10:19:59 PM asciidoctor
WARNING: listing with only inline code found; if this is not source from a file, consider using a literal block without source style instead; first line: maximum of two numbers [a: 1, b: 3, c: 3, #0]   PASSED
Mar 17, 2024 10:19:59 PM asciidoctor
WARNING: listing with only inline code found; if this is not source from a file, consider using a literal block without source style instead; first line: maximum of two numbers   FAILED
Mar 17, 2024 10:19:59 PM asciidoctor
WARNING: listing with only inline code found; if this is not source from a file, consider using a literal block without source style instead; first line: unroll {
```

leonard84 commented:
I think that there is a bit too much space added when the GitHub link is visible.
Furthermore, the icon top doesn't align with the box of the code listing.
The table also adds some padding to the left.

I tried to just insert a `<div style="float: right; margin-left:3px; max-width: 1em;">...` with the icons as the first element in the `content` div. I don't really know why the fa icons are so much bigger here than in the table though 🤔.

<img width="1030" alt="image" src="https://github.com/spockframework/spock/assets/1151511/ae4a2816-4e79-44ab-a5fa-59dbfa3b7291">

With `float: right; margin-left:3px; margin-top:1em; max-width: 1em;` it doesn't align with the top, but it doesn't overlap the `groovy` text. Not sure I like it better than the other.

<img width="1032" alt="image" src="https://github.com/spockframework/spock/assets/1151511/7f562f67-574e-4f8a-a792-0562ae4aea9b">

Vampire commented:
> When running :asciidoctor with these changes, I get a lot of these errors:

They are warnings, not errors.
And yes, that is intentional.
Our goal is, to have all code examples included from test files, isn't it?
These warnings warn about places where this is not the case.
That means the code is not verified by tests and there will also neither be a play linke, nor a GitHub link.
And if it is intentional at a place because it is not source input for example, a literal block can be used instead of a listing block.

Actually, the "without source style" was wrong since the cloaked literal block can and is be checked now, so it is removed now.

> I think that there is a bit too much space added when the GitHub link is visible.

I just create a table in AsciiDoc and let the layout be done by Asciidoctor / HTML / CSS.
I did not want to mess with the produced HTML.

> Furthermore, the icon top doesn't align with the box of the code listing.

Yes, that's intentional.
Icon 1x is a bit small.
Icon 2x is a bit large.
Icon 2x in superscript has a nice sice, and I personally like that it is a bit raised over the code block.

![image](https://github.com/spockframework/spock/assets/325196/cfcd5472-7af3-4858-90ab-6757f38ec69f)
<details>
  <summary>Here the plain source for these to play with</summary>

  ```adoc
  [source,groovy]
  ----
  runner {
    filterStackTrace false
  }
  ----
  
  [cols="a,0",frame=none,grid=none]
  |===
  |
  [source,groovy]
  ----
  runner {
    filterStackTrace false
  }
  ----
  |
  ^icon:github[2x,link=https://github.com,window=_blank]^
  ^icon:play[2x,link=https://github.com,window=_blank]^
  |===
  
  [cols="a,0",frame=none,grid=none]
  |===
  |
  [source,groovy]
  ----
  runner {
    filterStackTrace false
  }
  ----
  |
  icon:github[2x,link=https://github.com,window=_blank]
  icon:play[2x,link=https://github.com,window=_blank]
  |===
  
  [cols="a,0",frame=none,grid=none]
  |===
  |
  [source,groovy]
  ----
  runner {
    filterStackTrace false
  }
  ----
  |
  icon:github[link=https://github.com,window=_blank]
  icon:play[link=https://github.com,window=_blank]
  |===
  ```
</details>

> The table also adds some padding to the left.

So?
If you really have a problem with that, I guess you can adapt the used CSS.
We can easily give those tables an additional CSS class by assigning an AsciiDoc role like `.foo`.
Then in the CSS you should be able to change the padding settings for `.foo > tr > td` to your liking.
But I never was or will be a web designer.

> I tried to just insert a <div style="float: right; margin-left:3px; max-width: 1em;">... with the icons as the first element in the content div. I don't really know why the fa icons are so much bigger here than in the table though 🤔.

Probably because you do not use superscript which makes them a bit smaller, see above screenshot in this comment.

> With float: right; margin-left:3px; margin-top:1em; max-width: 1em; it doesn't align with the top, but it doesn't overlap the groovy text. Not sure I like it better than the other.

Well, as I said, I'm no web designer.
And I also did neither want to fiddle with the HTML, nor the CSS, but simply create some AsciiDoc that serves well. :-)

Vampire commented:
I configured the padding of the left cell to 0 now, that might result in what you like:
![image](https://github.com/spockframework/spock/assets/325196/d128ef4b-35b2-481c-b97d-4806911d5043)

glaforge commented:
FYI, I've just bought a domain name for the Groovy console, you'll be able to reference scripts hosted at:
https://groovyconsole.dev/

leonard84 commented on build-logic/asciidoc-extensions/src/main/java/org/spockframework/plugins/asciidoctor/IncludedSourceLinker.java:
```diff
@@ -0,0 +1,706 @@
+              // otherwise show the play icon
+              // but if the source link contains `/master/` instead of an commit ID,
+              // do not link the play icon, as the gwc cannot handle branch names
+              if (listings && !cloakedLiteral && parseBoolean(block.getAttribute("play", "true").toString())) {
```
I would prefer either `executable` or `webconsole`, `play` suggests some media player.

leonard84 reviewed (commented):
Just the rename of the `play` then this could be merged. I assume that the disabling of the execution will happen in another PR.

Vampire commented:
> I assume that the disabling of the execution will happen in another PR.

Imho, the disabling should never be used, and I did not plan to manually disable any.  And I surely do not have the time to look through each and every snippet's file to decide whether the snippet is expected to be executable or not.

This will also change over time, especially as you cannot choose with which Spock version the GWC should run.
